src/plugins/GLesmos/shaders.ts

```diff
@@ -0,0 +1,423 @@
+  }
+}
+
+const shaderCache = new Map<string, GLesmosProgram>();
+function getShaderProgram(
+  gl: WebGL2RenderingContext,
+  id: string,
+  vertexSource: string,
+  fragmentSource: string
+) {
+  const key = vertexSource + fragmentSource; // TODO: this is terrible
+  const cachedShader = shaderCache.get(key);
+
+  if (cachedShader) return cachedShader;
+
+  const shaderProgram = buildShaderProgram(
+    gl,
+    vertexSource,
+    fragmentSource,
+    id
+  ) as GLesmosProgram;
+
+  shaderProgram.vertexAttribPos = gl.getAttribLocation(
+    shaderProgram,
+    "vertexPosition"
+  );
+  shaderProgram.corner = gl.getUniformLocation(shaderProgram, "graphCorner");
+  shaderProgram.size = gl.getUniformLocation(shaderProgram, "graphSize");
+  shaderProgram.NaN = gl.getUniformLocation(shaderProgram, "NaN");
+  shaderProgram.Infinity = gl.getUniformLocation(shaderProgram, "Infinity");
+
+  shaderCache.set(key, shaderProgram);
+  if (shaderCache.size > 100) {
+    const key = Array.from(shaderCache.keys())[0];
+    shaderCache.delete(key);
+  }
+
+  return shaderProgram;
+}
+
+export const VERTEX_SHADER = `#version 300 es
+in highp vec2 vertexPosition;
+out vec2 texCoord;
+
+void main() {
+  texCoord    = vertexPosition * 0.5 + 0.5;
+  gl_Position = vec4(vertexPosition, 0.0, 1.0);
+}
+`;
+
+export const GLESMOS_ENVIRONMENT = `#version 300 es
+precision highp float;
+in  vec2 texCoord;
+out vec4 outColor;
+
+uniform vec2  graphCorner;
+uniform vec2  graphSize;
+uniform float NaN;
+uniform float Infinity;
+
+#define M_PI 3.1415926535897932384626433832795
+#define M_E 2.71828182845904523536028747135266
+
+vec2 toMathCoord(in vec2 fragCoord){
+  return fragCoord * graphSize + graphCorner;
+}
+
+vec4 mixColor(vec4 from, vec4 top) {
+  float a = 1.0 - (1.0 - from.a) * (1.0 - top.a);
+  return vec4((from.rgb * from.a * (1.0 - top.a) + top.rgb * top.a) / a, a);
+}
+`;
+
+export const GLESMOS_SHARED = `
+  vec4 getPixel( in vec2 coord, in sampler2D channel ){
+    return texture( channel, coord );
+  }
+
+  float line_segment(in vec2 p, in vec2 a, in vec2 b) {
+    vec2 ba = b - a;
+    vec2 pa = p - a;
+    float h = clamp(dot(pa, ba) / dot(ba, ba), 0., 1.);
+    return length(pa - h * ba);
+  }
+
+  float LineSDF(in vec4 line, in vec2 p){
+    return line_segment(p, vec2(line[0], line[1]), vec2(line[2], line[3]) );
+  }
+`;
+
+// = ===================== WebGL Source Generators ======================
+
+export function glesmosGetCacheShader(
+  gl: WebGL2RenderingContext,
+  id: string,
+  chunk: GLesmosShaderChunk,
+  deps: string
+): GLesmosProgram {
+  const source = `${GLESMOS_ENVIRONMENT}
+    // dependencies
+    ${deps}
+
+    // main func
+    ${chunk.def}
+
+    void main(){
+      vec2 mathCoord = texCoord * graphSize + graphCorner;
+      float v = _f0( mathCoord.x, mathCoord.y );
+      outColor = vec4(v, 0, 0, 1);
+    }
+  `;
+
+  const shader = getShaderProgram(gl, id, VERTEX_SHADER, source);
+
+  // TODO: set some uniforms here
```

Comment:
Is this needed?

Reply:
Since `getShaderProgram` returns an already-built shader if nothing has changed, it would be ideal to offload as many changes as possible into uniforms so we don't compile as much.

This comment is just here to remind someone (me) to do that later, and that it should be done here.  It would likely be a separate PR that packs all function variables into uniforms or something.